### PR TITLE
Return an empty collection, instead of 500 server error, for providers that don't support configured_systems or configuration_profiles subcollections

### PR DESCRIPTION
**Pull request by gtanzillo**
Before:
`curl --user admin:smartvm http://localhost:3000/api/providers/1000000000004/configuration_profiles -k`
```json
{"error":{"kind":"internal_server_error","message":"undefined method `configuration_profiles' for #\u003cManageIQ::Providers::Google::CloudManager:0x000055ba49eb1168\u003e","klass":"NoMethodError"}}  
```
`curl --user admin:smartvm http://localhost:3000/api/providers/1000000000004/configured_systems -k `
```json
{"error":{"kind":"internal_server_error","message":"undefined method `configured_systems' for #\u003cManageIQ::Providers::Google::CloudManager:0x000055b866e0e178\u003e","klass":"NoMethodError"}} 
```
After:
```json
curl --user admin:smartvm http://localhost:3000/api/providers/1000000000004/configuration_profiles
{
    "name": "configuration_profiles",
    "count": 0,
    "subcount": 0,
    "pages": 0,
    "resources": [],
    "links": {
        "self": "http://localhost:3000/api/providers/1000000000004/configuration_profiles?offset=0",
        "first": "http://localhost:3000/api/providers/1000000000004/configuration_profiles?offset=0",
        "last": "http://localhost:3000/api/providers/1000000000004/configuration_profiles?offset=0"
    }
}
```
```json
curl --user admin:smartvm http://localhost:3000/api/providers/1000000000004/configured_systems
{
    "name": "configured_systems",
    "count": 143,
    "subcount": 0,
    "pages": 1,
    "resources": [],
    "links": {
        "self": "http://localhost:3000/api/providers/1000000000004/configured_systems?offset=0",
        "first": "http://localhost:3000/api/providers/1000000000004/configured_systems?offset=0",
        "last": "http://localhost:3000/api/providers/1000000000004/configured_systems?offset=0"
    }
}
```
/cc @lpichler

### CODE REVIEW

**Comment by lpichler** [comment]
Relations `configured_systems` and `configuration_profiles` belong to just AutomationManagers and ConfigurationManagers and so it can be reason why relations should stay here. I guess such type of issue can happen only with `/providers` endpoint.

Question:

How to tackle queries on subcollections:

```
GET /api/providers/?expand=resources&attributes=configuration_profiles

{
    "error": {
        "kind": "bad_request",
        "message": "Invalid attributes specified: configuration_profiles",
        "klass": "Api::BadRequestError"
    }
}
```

but 

```
GET /api/providers/?expand=resources&attributes=configuration_profiles&filter[]=id=1

{
    "name": "providers",
    "count": 4,
    "subcount": 1,
    "subquery_count": 1,
    "pages": 1,
    "resources": [
        {
            "href": "http://localhost:3090/api/providers/1",
            "id": "1",
            "name": "Embedded Ansible",
            "created_on": "2020-07-10T13:20:27Z",
            "updated_on": "2020-07-10T13:27:33Z",
            "guid": "e9b5f3fd-c5f0-4d2d-a2cd-0843e656be1f",
            "zone_id": "2",
            "type": "ManageIQ::Providers::EmbeddedAnsible::AutomationManager",
            "api_version": null,
            "uid_ems": null,
            "host_default_vnc_port_start": null,
            "host_default_vnc_port_end": null,
            "provider_region": null,
            "last_refresh_error": "uninitialized constant ManageIQ::Providers::EmbeddedAnsible::Inventory",
            "last_refresh_date": "2020-07-10T13:27:33Z",
            "provider_id": "1",
            "realm": null,
            "tenant_id": "1",
            "project": null,
            "parent_ems_id": null,
            "subscription": null,
            "last_metrics_error": null,
            "last_metrics_update_date": null,
            "last_metrics_success_date": null,
            "tenant_mapping_enabled": null,
            "enabled": true,
            "options": null,
            "zone_before_pause_id": null,
            "last_inventory_date": null,
            "configuration_profiles": [],
            ...
```


so I am thinking whether is better to return just `[]` or `nil`. What do think about it?

**Comment by bdunne** [comment]
I think it's important to maintain a consistent interface.  So, I think it would be better to return `[]` rather than `nil`.  I think we could do that by moving the relation to the base class.

**Comment by agrare** [comment]
That would fix this subcollection, but there are _lots_ of associations that only exist on certain types of providers.  For example asking for any of [these](https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/cloud_manager.rb#L17-L29) off of anything other than a CloudManager will fail in the same way.  I don't know that it is feasible to move every association from every provider sub-type into the base class

**Comment by agrare** [comment]
> For the other case, where this endpoint returns Providers rather than ExtManagementSystems, I don't know what to do with that. At some point we need to fix this interface.

Yeah the `/providers` endpoint is a mess

**Comment by gtanzillo** [comment]
Based on @agrare comments, it seems like the safest change is to the subcollections. But I'll change to return an empty list instead of a 404, as suggested.

**Comment by abellotti** [comment]
empty list is the right way, as the subcollection endpoint does exist,   other subcollections do the same,

i.e.  ```object.try(:cloud_networks) || []```. in the *_query_resource method.

File: spec/requests/providers_spec.rb [hunk 1]
```diff
@@ -1875,6 +1887,16 @@ class DummyProvider; end
 
         expect(response).to have_http_status(:forbidden)
       end
+
+      it "returns a 404 for a provider that does not support configured_systems" do
```
Comment (abellotti) [inline spec/requests/providers_spec.rb:1891]: minor, maybe update this and the PR description.

Comment (gtanzillo) [reply spec/requests/providers_spec.rb:1891]: Definitely, I meant to update the test case descriptions and forgot. Though I updated the PR description but I must have forgotten to save. I'll take care of both. Thanks!

**Comment by abellotti** [comment]
LGTM !!  Thanks @gtanzillo  for fixing this. 👍

**Comment by simaishi** [comment]
Jansa backport details:

```
$ git log -1
commit 30bcf0c130d9f87274aae113056bda3c70dff529
Author: Alberto Bellotti <abellotti@users.noreply.github.com>
Date:   Wed Jul 15 08:11:48 2020 -0400

    Merge pull request #864 from gtanzillo/return-404

    Return an empty collection, instead of 500 server error, for providers that don't support configured_systems or configuration_profiles subcollections

    (cherry picked from commit e8db32cb0995bd733888f45f36f41a00dc80ae8b)
```

**Pull request merged**